### ISSUE TITLE
Segfault (related to notation printing?)

### ISSUE THREAD

**Issue by Janno** [coq/coq#6821]
#### Version
8.7.1

#### Operating system
Linux

#### Description of the problem

The following code makes coqtop segfault.

```Coq
Notation "x 'where' y .. z := v " :=
  ((fun y => .. ((fun z => x) v) ..) v) (at level 0, v at next level, y binder, z binder).

Check forall f, f x where x := 1.
```

Note that the following `Check` invocation does not segfault:
```Coq
Check x where x := 1.
```

**Comment by anton-trunov** [comment]
macOS 10.13.3:

```
Coq < Notation "x 'where' y .. z := v " := ((fun y => .. ((fun z => x) v) ..) v) (at level 0, v at next level, y binder, z binder).

Coq < Check forall f, f x where x := 1.
Toplevel input, characters 0-33:
> Check forall f, f x where x := 1.
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Stack overflow.
```

**Comment by Zimmi48** [comment]
I get a stack overflow as well with coqtop 8.7.1 on Linux (NixOS).

**Comment by herbelin** [comment]
Well, the problem was in the printing of recursive notation. A short example was:
```
Notation "x 'where' y .. z := v " :=
  ((fun y => .. ((fun z => x) v) ..) v) (at level 0, v at next level, y binder, z binder).
Check fun f => f 0. (* Stack overflow *)
```
The term `f` was eta-expanded to match the pattern, but this eta-expansion was leading to the same problem recursively.

Fixed in #6823.
